### PR TITLE
Fix compilation error with generic type attributes

### LINKED ISSUE #97498: Logging source generator fails to compile if class has [DynamicallyAccessedMembers]
### Description

If the logging source generator is used in a .NET 8 application that is intended to be native AoT compiled and is used on a class with a generic type that is annotated with a `[DynamicallyAccessedMembers]` attribute, the project fails to compile.

```
Microsoft.Extensions.Logging.Generators\Microsoft.Extensions.Logging.Generators.LoggerMessageGenerator\LoggerMessage.g.cs(4,31): error CS0246: The type or namespace name 'DynamicallyAccessedMembersAttribute' could not be found (are you missing a using directive or an assembly reference?)
Microsoft.Extensions.Logging.Generators\Microsoft.Extensions.Logging.Generators.LoggerMessageGenerator\LoggerMessage.g.cs(4,31): error CS0246: The type or namespace name 'DynamicallyAccessedMembers' could not be found (are you missing a using directive or an assembly reference?)
Microsoft.Extensions.Logging.Generators\Microsoft.Extensions.Logging.Generators.LoggerMessageGenerator\LoggerMessage.g.cs(4,58): error CS0103: The name 'DynamicallyAccessedMemberTypes' does not exist in the current context
```

Without the attribute, `dotnet publish` produces a warning:

```
Trim analysis warning IL2075: Describer`1.Describe(!0): 'this' argument does not satisfy 'DynamicallyAccessedMemberTypes.PublicProperties' in call to 'System.Type.GetProperties()'. The return value of method 'System.Object.GetType()' does not have matching annotations. The source value must declare at least the same requirements as those declared on the target location it is assigned to.
```

### Reproduction Steps

Clone [this repository](https://github.com/martincostello/logging-source-generator-dynamically-accessed-members) repository and run `dotnet build` from the root directory.

### Expected behavior

- The project compiles.
- `dotnet publish` succeeds with no warnings.

### Actual behavior

The project fails to compile with `CS0246` and `CS0103` errors.

### Regression?

Unknown - I've only tried to do this code pattern once, and this was the result.

### Known Workarounds

Suppress the generated `IL2075` in code in a simple program tested to otherwise work.

This is not likely to be appropriate for a larger application that cannot be otherwise verified to be safe.

### Configuration

```console
❯ dotnet --info
.NET SDK:
 Version:           8.0.101
 Commit:            6eceda187b
 Workload version:  8.0.100-manifests.66f5ce51

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.22621
 OS Platform: Windows
 RID:         win-x64
 Base Path:   C:\Program Files\dotnet\sdk\8.0.101\

.NET workloads installed:
 Workload version: 8.0.100-manifests.66f5ce51
There are no installed workloads to display.

Host:
  Version:      8.0.1
  Architecture: x64
  Commit:       bf5e279d92

.NET SDKs installed:
  6.0.321 [C:\Program Files\dotnet\sdk]
  6.0.418 [C:\Program Files\dotnet\sdk]
  7.0.312 [C:\Program Files\dotnet\sdk]
  7.0.405 [C:\Program Files\dotnet\sdk]
  8.0.101 [C:\Program Files\dotnet\sdk]
  8.0.200-preview.23624.5 [C:\Program Files\dotnet\sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.App 6.0.25 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.26 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 7.0.14 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 7.0.15 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 8.0.0 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 8.0.1 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 6.0.25 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.26 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 7.0.14 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 7.0.15 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 8.0.0 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 8.0.1 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 6.0.25 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 6.0.26 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 7.0.14 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 7.0.15 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 8.0.0 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 8.0.1 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]

Other architectures found:
  x86   [C:\Program Files (x86)\dotnet] registered at [HKLM\SOFTWARE\dotnet\Setup\InstalledVersions\x86\InstallLocation]

Environment variables:
  Not set

global.json file:
  C:\Coding\martincostello\logging-source-generator-dynamically-accessed-members\global.json

Learn more:
  https://aka.ms/dotnet/info

Download .NET:
  https://aka.ms/dotnet/download
```

### Other information

_No response_

### PR DESCRIPTION
**Pull request by martincostello**
Strip generic type parameter attributes from partial classes emitted by the logging source generator to avoid CS0579 errors from duplicate attributes.

Fixes #97498.

### CODE REVIEW

**Comment by tarekgh** [comment]
@martincostello The original issue reported in https://github.com/dotnet/runtime/issues/97498 was about `The type or namespace name 'DynamicallyAccessedMembersAttribute' could not be found (are you missing a using directive or an assembly reference?)`. But you are saying `Strip generic type parameter attributes from partial classes emitted by the logging source generator to avoid CS0579 errors from duplicate attributes.` . 

I expect to fix this by fully qualifying the attribute in the emitted source. What exactly causes `CS0579 errors from duplicate attributes`?

**Comment by tarekgh** [comment]
CC @eerhardt if you have any input for this.

**Comment by martincostello** [comment]
> What exactly causes `CS0579 errors from duplicate attributes`?

Having the attributes defined in both partial classes, it just flat-out doesn't compile. Adding the usings/fully-qualifying just causes a new CS0579 error.

https://github.com/dotnet/runtime/issues/97498#issuecomment-1913599160

As far as I can tell, only one partial class can annotate the type parameters, so the generator shouldn't emit them and leave it up to the user's code to do so. That's what this PR does.

**Comment by tarekgh** [comment]
CC @geeknoid and @joperezr as FYI.

**Comment by tarekgh** [comment]
@martincostello I see what you are saying now. Thanks for clarification.

File: src/libraries/Microsoft.Extensions.Logging.Abstractions/gen/LoggerMessageGenerator.Parser.cs
Comment (tarekgh) [inline src/libraries/Microsoft.Extensions.Logging.Abstractions/gen/LoggerMessageGenerator.Parser.cs:610]: Could you please change this comment to something like:
```
The source generator produces a partial class that the compiler merges with the original class definition in the user code. If the user applies attributes to the generic types of the class, it is necessary to remove these attribute annotations from the generated code. Failure to do so may result in a compilation error (CS0579: Duplicate attribute).
```

File: src/libraries/Microsoft.Extensions.Logging.Abstractions/gen/LoggerMessageGenerator.Parser.cs
Comment (tarekgh) [inline src/libraries/Microsoft.Extensions.Logging.Abstractions/gen/LoggerMessageGenerator.Parser.cs:617]: >typeDeclaration = typeDeclaration.ReplaceNode(parameter, parameter.WithAttributeLists([])); [](http://example.com/codeflow?start=28&length=91)

Have you verified whether the final IL code generated by the compiler still contains the attribute? While we are addressing the issue in the generated code, it is crucial to confirm that this correction does not unintentionally remove the attribute from the combined partial classes emitted by the compiler.

File: src/libraries/Microsoft.Extensions.Logging.Abstractions/tests/Microsoft.Extensions.Logging.Generators.Tests/LoggerMessageGeneratorEmitterTests.cs
Comment (tarekgh) [inline src/libraries/Microsoft.Extensions.Logging.Abstractions/tests/Microsoft.Extensions.Logging.Generators.Tests/LoggerMessageGeneratorEmitterTests.cs:166]: >TestBaseline_TestWithNestedClassWithGenericTypesWithAttributes_Success [](http://example.com/codeflow?start=26&length=70)

could you please add some test using the reflection to ensure the class GenericTypeWithAttribute type parameters still has the attributes Foo, Bar?

File: src/libraries/Microsoft.Extensions.Logging.Abstractions/gen/LoggerMessageGenerator.Parser.cs
Comment (martincostello) [reply src/libraries/Microsoft.Extensions.Logging.Abstractions/gen/LoggerMessageGenerator.Parser.cs:617]: I hadn't as it didn't occur to me that this could be possible as my understanding is that it's editing the model which is then compiled for the source emitted, rather than changing the total sum of the source for the target type from all its sources.

I've just checked, and the attributes are still present if I view the compiled test DLL in ILSpy:

image

**Review by tarekgh — APPROVED** [review]
Thanks @martincostello for your help providing the fix!

**Comment by martincostello** [comment]
@tarekgh No problem!

Is this change something that could be considered for backporting to 8.0 after merge?

We have a base library targeting `net8.0` we use internally for building serverless apps that makes heavy use of generics and the logging source generator. I found this issue while working on adding AoT support to it.

This issue blocks us from adopting native AoT there without significant refactoring of all the logging in that library, which in turn blocks all the apps built on top of it.

**Comment by tarekgh** [comment]
> Is this change something that could be considered for backporting to 8.0 after merge?

I can check if this is meeting the servicing bar. Let's get this first in the main and then we can follow-up to see if we can port it to servicing.

CC @ericstj

**Review by geeknoid — APPROVED** [review]

**Comment by tarekgh** [comment]
@martincostello this is now ported to servicing branch. Will be released around March.

**Comment by martincostello** [comment]
@tarekgh Thanks! ❤️

**Pull request merged**
